PKCS12.xs

```diff
@@ -325,19 +519,437 @@ int dump_certs_keys_p12(BIO *bio, PKCS12 *p12, char *pass, int passlen, int opti
     }
 
+  if (orig_hash) {
+    if((hv_store(orig_hash, "pkcs7_data", strlen("pkcs7_data"), newRV_inc((SV *) pkcs7_bags_av), 0)) == NULL)
+      croak("unable to add bags to the hash");
+    if((hv_store(orig_hash, "pkcs7_encrypted_data", strlen("pkcs7_encrypted_data"), newRV_inc((SV *) pkcs7_enc_bags_av), 0)) == NULL)
+      croak("unable to add bags to the hash");
+  }
     sk_PKCS12_SAFEBAG_pop_free(bags, PKCS12_SAFEBAG_free);
   }
 
   sk_PKCS7_pop_free(asafes, PKCS7_free);
 
   return 1;
 }
+#ifdef TIM
+# define B_FORMAT_TEXT   0x8000
+# define FORMAT_TEXT    (1 | B_FORMAT_TEXT)     /* Generic text */
+int FMT_istext(int format)
+{
+  return (format & B_FORMAT_TEXT) == B_FORMAT_TEXT;
+}
+
+BIO *dup_bio_err(int format)
+{
+  BIO *b = BIO_new_fp(stderr,
+                      BIO_NOCLOSE | (FMT_istext(format) ? BIO_FP_TEXT : 0));
+
+  return b;
+}
+#endif
+static unsigned long nmflag = 0;
+static char nmflag_set = 0;
+# define XN_FLAG_SPC_EQ          (1 << 23)/* Put spaces round '=' */
+
+#define CHARTYPE_BS_ESC         (ASN1_STRFLGS_ESC_2253 | CHARTYPE_FIRST_ESC_2253 | CHARTYPE_LAST_ESC_2253)
+
+#define ESC_FLAGS (ASN1_STRFLGS_ESC_2253 | \
+                   ASN1_STRFLGS_ESC_QUOTE | \
+                   ASN1_STRFLGS_ESC_CTRL | \
+                   ASN1_STRFLGS_ESC_MSB)
+
+unsigned long get_nameopt(void)
+{
+  return
+      nmflag_set ? nmflag : ESC_FLAGS | XN_FLAG_SEP_CPLUS_SPC | ASN1_STRFLGS_UTF8_CONVERT | XN_FLAG_SPC_EQ;
+}
+
+void print_name(BIO *out, const char *title, CONST_X509_NAME *nm)
+{
+  char *buf;
+  char mline = 0;
+  int indent = 0;
+  unsigned long lflags = get_nameopt();
+
+  if (out == NULL)
+    return;
+  if (title != NULL)
+    BIO_puts(out, title);
+  if ((lflags & XN_FLAG_SEP_MASK) == XN_FLAG_SEP_MULTILINE) {
+    mline = 1;
+    indent = 4;
+  }
+  if (lflags == XN_FLAG_COMPAT) {
+    buf = X509_NAME_oneline(nm, 0, 0);
+    BIO_puts(out, buf);
+    BIO_puts(out, "\n");
+    OPENSSL_free(buf);
+  } else {
+    if (mline)
+      BIO_puts(out, "\n");
+      X509_NAME_print_ex(out, nm, indent, lflags);
+
+      BIO_puts(out, "\n");
+  }
+}
+
+void dump_cert_text(BIO *out, X509 *x)
+{
+  print_name(out, "subject=", X509_get_subject_name(x));
+  print_name(out, "issuer=", X509_get_issuer_name(x));
+}
+
+SV * get_cert_subject_name(pTHX_ X509 *x)
+{
+  BIO *bio;
+  CHECK_OPEN_SSL(bio = BIO_new(BIO_s_mem()));
+
+  print_name(bio, "", X509_get_subject_name(x));
+  SV *sv = extractBioString(aTHX_ bio);
+  return sv;
+}
+
+SV * get_cert_issuer_name(pTHX_ X509 *x)
+{
+  BIO *bio;
+  CHECK_OPEN_SSL(bio = BIO_new(BIO_s_mem()));
+
+  print_name(bio, "", X509_get_issuer_name(x));
+  SV *sv = extractBioString(aTHX_ bio);
+  return sv;
+}
+
+void get_hex(char *out, unsigned char *buf, int len)
+{
+  int i;
+  for (i = 0; i < len; i++) {
+    if (i == (len-1))
+      out += sprintf(out, "%02X", buf[i]);
+    else
+      out += sprintf(out, "%02X ", buf[i]);
+  }
+}
+
+void hex_prin(BIO *out, unsigned char *buf, int len)
```

Comment:
Hi @timlegge 

Was this method to be named: `hex_print` ?

Reply:
That is a copy paste from openssl I believe.  https://github.com/openssl/openssl/blob/0cd9dd703ea575699b2d3cd74f1b8224447f4352/apps/pkcs12.c#L1320